use_case_guide/real_time_dashboards.rst

```diff
@@ -138,141 +154,77 @@ and Citus will place matching shards on the same worker. This is called
-        ) SELECT
-          site_id,
-          date_trunc('minute', ingest_time) as minute,
-          COUNT(1) as request_count,
-          SUM(CASE WHEN (status_code between 200 and 299) THEN 1 ELSE 0 END) as success_count,
-          SUM(CASE WHEN (status_code between 200 and 299) THEN 0 ELSE 1 END) as error_count,
-          SUM(response_time_msec) / COUNT(1) AS average_response_time_msec
-        FROM %1$I
-        WHERE
-          date_trunc('minute', ingest_time)
-            > (SELECT COALESCE(max(ingest_time), timestamp '10-10-1901') FROM %2$I)
-          AND date_trunc('minute', ingest_time) < date_trunc('minute', now())
-        GROUP BY site_id, minute
-        ORDER BY minute ASC;
-      $insert$, p_source_shard, p_dest_shard);
-    END;
-    $$ LANGUAGE 'plpgsql';
-
-Inside this function you can see the dashboard query from earlier. It's been wrapped in
-some machinery which writes the results into ``http_request_1min`` and allows passing in
-the name of the shards to read and write from. It also takes out an advisory lock, to
-ensure there aren't any concurrency bugs where the same rows are written multiple times.
-
-The machinery above which accepts the names of the shards to read and write is necessary
-because only the coordinator has the metadata required to know what the shard pairs are. It has
-its own function to figure that out:
+In order to populate ``http_request_1min`` we're going to periodically run
+an INSERT INTO SELECT. This is possible because the tables are co-located.
+The following function wraps the rollup query up for convenience.
 
 .. code-block:: plpgsql
 
-    -- this function is created on the coordinator
-    CREATE FUNCTION colocated_shard_placements(left_table REGCLASS, right_table REGCLASS)
-    RETURNS TABLE (left_shard TEXT, right_shard TEXT, nodename TEXT, nodeport INT4) AS $$
-      SELECT
-        a.logicalrelid::regclass||'_'||a.shardid,
-        b.logicalrelid::regclass||'_'||b.shardid,
-        nodename, nodeport
-      FROM pg_dist_shard a
-      JOIN pg_dist_shard b USING (shardminvalue)
-      JOIN pg_dist_placement p ON (a.shardid = p.shardid)
-      JOIN pg_dist_node n ON (p.groupid = n.groupid)
-      WHERE a.logicalrelid = left_table AND b.logicalrelid = right_table;
-    $$ LANGUAGE 'sql';
-
-Using that metadata, every minute it runs a script which calls ``rollup_1min`` once for
-each pair of shards:
-
-.. code-block:: bash
-
-   #!/usr/bin/env bash
-
-   QUERY=$(cat <<END
-     SELECT * FROM colocated_shard_placements(
-       'http_request'::regclass, 'http_request_1min'::regclass
-     );
-   END
-   )
-
-   COMMAND="psql -h \$2 -p \$3 -c \"SELECT rollup_1min('\$0', '\$1')\""
-
-   psql -tA -F" " -c "$QUERY" | xargs -P32 -n4 sh -c "$COMMAND"
+  CREATE OR REPLACE FUNCTION rollup_http_request() RETURNS void AS $$
+  DECLARE max_rollup_time TIMESTAMPTZ;
+  BEGIN
+    SELECT COALESCE(max(ingest_time), timestamp '10-10-1901')
```

Comment:
If I take the insert…select out of the function and use a subquery directly (rather than `max_rollup_time`)

```sql
WHERE minute > (
    SELECT COALESCE(max(ingest_time), timestamp '10-10-1901')
    FROM http_request_1min
  )
  AND minute < date_trunc('minute', now())
```

Then it gives me an error

```
ERROR:  0A000: cannot perform distributed planning for the given modification
DETAIL:  Select query cannot be pushed down to the worker.
LOCATION:  RouterModifyTaskForShardInterval, insert_select_planner.c:548
```

I think it's because we're trying to find the max over the whole `http_request_1min` table which is inaccessible from down in a worker.

Reply:
Just tried another approach: calculating the max for each tenant rather than globally. Should still process only the new data on each tenant.

```sql
  WHERE minute > (
      SELECT COALESCE(max(ingest_time), timestamp '10-10-1901')
      FROM http_request_1min
      WHERE http_request_1min.site_id = h.site_id
    )
    AND minute < date_trunc('minute', now())
```

This allows the insert…select query to run. What do you think, should I use this instead?

Reply:
As we discussed yesterday we are going ahead with leaving the rollup in a function. Then we don't need to simplify and make this one query. So we can retain what we have now.